Fixes not all post search results showing up sometimes

oguzkocer opened the pull request:
Fixes #11080. This PR utilizes the `PostSummary` that's introduced in https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1479 to section to post list. The problem with the previous implementation is that, during `PostListItemDataSource.getItemIdentifiers` the actual models might not be in the DB, so when we rely on `PostModelTable` we end up filtering out the search results. Instead, what we should do is to keep a separate table in the DB to keep the post statuses, so we can always look them up. Please see the FluxC PR for more details.

While working on this, I made some changes to `getGroupedItemIdentifiers` to simplify it. I also ended up re-writing how the unit tests work. In my opinion, these tests are setup in a way that makes it very hard to understand and maintain them. Specifically, I feel that the reliance on verifying specific ids without them being extracted to constants (which was really not a great option here) is almost never a good option. Furthermore, in this instance specifically, ids can be either `LocalId` or `RemoteId` and relying on a specific one can be very error prone depending on how the mocking is setup.

I believe we could use a parametrized tests for all this, because essentially we are always trying to verify the same thing. Given a set of identifiers and their status, we want to check whether we add the correct section headers to the list and whether the items are put under their sections correctly. I didn't go as far as to make this a parametrized test, but I've refactored the code in a way that is pretty much how it works now, a single test runner with multiple different scenarios.

_Note: I didn't make this parametrized right now, because all tests are under `PostListItemDataSourceTest` name which indicates more than just sectioning tests. I didn't want to change that too since I made a lot more changes in this PR than I initially intended to._

**To test:**

It's a little hard to describe exact testing steps for #11080, so I am going to explain what needs to be tested instead of a simple set of instructions. Please let me know if you need help with testing.

* Start with a fresh installation so there are no posts in the DB
* Visit the posts page for a REST API powered site (even though we have `PostSummary` support for self-hosted sites, there is no search functionality, so it's not possible to test the sectioning)
* Do not scroll at all
* Go to search and type a query that you know will give you results that are not in the DB yet. You can use Stetho to check which posts you have or use a very big blog and search something old or search something in Trash
* Verify that you get the results you'd get if you searched on Calypso and posts are in the correct section. If you briefly see a post loading, that means it's working because that animation only plays if the post is not in the DB.

Having said all that, if you use post search and verify that sections are working correctly, that'd be enough because #11080 can't happen anymore with these changes.

**Here are my tests:**
* Before the fix: https://cloudup.com/cSht5x-LKkd
* After the fix: https://cloudup.com/cgILLs6j88Y
Notice how many more results I get after the fix.

**PR submission checklist:**
- [x] I have considered adding unit tests where possible.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

cc @chipsnyder

chipsnyder commented:
> I scroll to the bottom of the "Drafts" section and new posts appear at the top of "Published" section.

@planarvoid and I discussed this as part of my trial. The only things we could come up with were to 
A) implement it the way it is
B) turn off pagination
C) Modify ListStore to accept a ListDiscriptor per section. 

I can go into more details on option C but we ultimately felt like that turned into a larger project in itself

oguzkocer commented:
It's true that we would want to run the existing tests before making any changes to it, but I never thought it'd be the reviewer that needs to do that. Your comment kind of sounds like we should not touch unit tests when we are making changes that unit test is verifying and I don't think I can agree with that. However, if you feel specifically for this PR it should be done that way, we can definitely do that.

---

I think we have a different take on what an easy to understand test means. I had the opposite experience when I was looking at the existing tests and had a hard time trying to understand what's going on especially with all the ids. I respect wanting simpler tests, but existing tests were not that for me. Of course, it might just be me.

---

For the new tests, first of all, I'd consider them integration tests rather than simple unit tests. As you mentioned, for test itself to be simple the code it's testing needs to be simple. Well, simple is a subjective word, so better put, the closer we are to pure functions the easier time we'll have testing it and understanding it.

This actually comes down to an argument we had in the Vegas meetup about being able to test private classes/functions. In Java/Kotlin, we keep having to make a choice between encapsulation and unit testability. Then we end up in these weird situations where a function has to do a lot of things at the same time (even if through simple private helpers) and us having to write unit tests that needs to behave as an integration test all so we can preserve encapsulation. If, for example, we had a helper function that returned a map instead of a list to separate statuses into sections with a parameter to check status of an item, it'd be much simpler to test that. Then a function that aggregates the results would be very easy to test as well.

---

In this PR, I considered extracing sectioning logic somewhere else (public), so we'll have much simpler functions, so we can verify smaller behaviour in these tests instead. However, that was not the goal of the PR, so changing that as well felt wrong. In hindsight, maybe it'd have been better with how the tests turned out, but I didn't see that complication from the beginning. To help with that, I updated the documentation for the test in 3fe759f215b47e53d9de3b7c7568922462b8af7b, so let me know how you feel about that.

If you feel that previous tests were better, I don't mind reverting to that. Just for my concience though I have to say I think that'd be a mistake. Not because I wrote it, because you know I scratch a lot of code I write if it doesn't turn out well, but because the new test is not only more maintainable but also a much safer way to test the whole functionality with much less chance to result in false positives & false negatives.

---

> The posts in search are ordered from the oldest to the newest (I'd expect them to be ordered other way round).

Good catch, I missed this! It's actually a result of this PR since `isIn` sqlite query will lose the order. I have fixed that in 8cd4b0378a0d25fbc65a647ae36455f5af77e82b which includes the fix in FluxC: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/1479/commits/948383d71745237fdaf4bb58051b62011446feaa

---

> I scroll to the bottom of the "Drafts" section and new posts appear at the top of "Published" section.

I think this is just a simple case of requesting the `/posts` endpoint to order the result first in terms of `status` then in terms of `date`. I don't think the endpoint is currently capable of this, but we should be able to request a change to it. I don't think any other solution in the client makes sense.

> C) Modify ListStore to accept a ListDiscriptor per section.

@chipsnyder I've explored this path before, the complexity it brings is definitely not worth it in my opinion. Especially since there is a much easier way to do it.

oguzkocer commented on WordPress/src/test/java/org/wordpress/android/viewmodel/posts/PostListItemDataSourceTest.kt:
```diff
@@ -1,172 +1,163 @@
+            val currentSection = mutableSectionHeaders.first()
+            when (identifier) {
+                is LocalPostId -> {
+                    assert(PostListType.fromPostStatus(getLocalPostStatus(identifier)) == currentSection.type)
```
I don't actually know what the difference is, but switched to `assertTrue` of junit in c91f76d4b53b546cb692ffa2233659d8b2a032b7. Let me know if that works for you.

malinajirka replied:
A couple of differences
-  `assert` works only when you start the JVM with a flag -> it always passes otherwise. (I thought the AS is smart enough to turn it on when running unit-tests. But it actually isn't ->  the tests were actually always passing no matter the value)
- `assertThat` automatically tells you what was the expected and actual values when the test fails (as oppose to `assertTrue` and `assert`) (more differences [here](https://objectpartners.com/2013/09/18/the-benefits-of-using-assertthat-over-other-assert-methods-in-unit-tests/))
- consistency:) - we use `assertThat` in most unit tests (`assertTrue` is used in over 100 places as well, so I'm fine with that.)

`assert` is sometimes for example used in debug builds to follow the fail fast principle.

oguzkocer replied:
@malinajirka Thank you for elaborating on this. I've made the change to `assertThat` in f0906511c5d50c6ac2ac0b2347c89484d4a63c8f. I should have done that instead of `assertTrue` in the first place, sorry about that!

malinajirka replied:
No worries at all, thanks! ;)

oguzkocer commented:
@malinajirka I've updated the FluxC hash, fixed the `assertThat` issue and merged `develop` into this branch to fix release notes conflict and marked the PR as ready for review.

> I scroll to the bottom of the "Drafts" section and new posts appear at the top of "Published" section.

To close the loop on this issue, I've made a request for the endpoint change.

malinajirka commented:
Thanks @oguzkocer!

I've found a bug, but I believe it's not related to this PR. I'll create a new ticket in FluxC.

oguzkocer commented:
Hey @jkmassel! I wanted to update the WordPressMocks library with the changes from this PR [as pinged by the bot](https://github.com/wordpress-mobile/WordPress-Android/pull/11098#issuecomment-575312596) however the files I want to change don't exist in `develop`. I do see them in the `master` branch, but since we generally use `develop`, I thought I'd check in before making a change. Should I open a PR against `master` or is there something wrong with `develop` that needs to be addressed?
